cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java

```diff
@@ -568,19 +593,147 @@ private void searchToolsInCentral(String keyword) {
+            CommandUtil.exitError(this.exitWhenFinish);
+            return;
+        }
+
+        Path balaCacheDirPath = ProjectUtils.createAndGetHomeReposPath()
+                .resolve(REPOSITORIES_DIR).resolve(CENTRAL_REPOSITORY_CACHE_NAME)
+                .resolve(ProjectConstants.BALA_DIR_NAME);
+
+        // TODO: make supportedPlatform = "java11,java17" once central is fixed
+        String supportedPlatform = "java11";
+        try {
+            version = getLatestVersionForUpdateCommand(tool.get());
+            if (tool.get().version().equals(version)) {
+                outStream.println("tool '" + toolId + "' is already up-to-date.");
+                CommandUtil.exitError(this.exitWhenFinish);
+                return;
+            }
+            if (isToolAvailableLocally(toolId, version)) {
+                outStream.println("tool " + toolId + ":" + version + " is already available locally.");
+            } else {
+                pullToolFromCentral(supportedPlatform, balaCacheDirPath);
+            }
+            addToBalToolsToml();
+        } catch (PackageAlreadyExistsException e) {
+            errStream.println(e.getMessage());
+            CommandUtil.exitError(this.exitWhenFinish);
+        } catch (CentralClientException | ProjectException e) {
+            CommandUtil.printError(errStream, "unexpected error occurred while pulling tool:" + e.getMessage(),
+                    null, false);
+            CommandUtil.exitError(this.exitWhenFinish);
+        }
+    }
+
+    private String getLatestVersionForUpdateCommand(BalToolsManifest.Tool tool) throws CentralClientException {
+        Settings settings;
+        try {
+            settings = RepoUtils.readSettings();
+            // Ignore Settings.toml diagnostics in the pull command
+        } catch (SettingsTomlException e) {
+            // Ignore 'Settings.toml' parsing errors and return empty Settings object
+            settings = Settings.from();
+        }
+        System.setProperty(CentralClientConstants.ENABLE_OUTPUT_STREAM, "true");
+        CentralAPIClient client = new CentralAPIClient(RepoUtils.getRemoteRepoURL(),
+                initializeProxy(settings.getProxy()), settings.getProxy().username(),
+                settings.getProxy().password(), getAccessTokenOfCLI(settings));
+        List<String> versions = client.getPackageVersions(tool.org(), tool.name(), ANY_PLATFORM,
+                RepoUtils.getBallerinaVersion());
+        return getLatestVersionInSameMinor(versions, tool.version());
+    }
+
+    private String getLatestVersionInSameMinor(List<String> versions, String currentVersionStr) {
+        SemanticVersion currentVersion = SemanticVersion.from(currentVersionStr);
+        Optional<String> latestVersionInSameMinor = versions.stream().map(SemanticVersion::from)
+                .filter(version -> version.major() == currentVersion.major()
```

Comment:
The latest version of the tool compatible with the current distribution should be downloaded and set as the active version.

Reply:
We are passing a header named `User-Agent` that contains our distribution version and we will not get any versions that are built with newer updates from the endpoint `/packages/<org_name>/<tool_id>`. Therefore the `latestVersionInSameMinor` will be always compatible.

Reply:
Changed the logic to pull the latest compatible. (including major, minor, patch)